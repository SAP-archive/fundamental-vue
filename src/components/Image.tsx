import { Component, Prop, Base } from '@/core';

const sizeMapping = {
  s: 'Small',
  m: 'Medium (default)',
  l: 'Large',
};
type IdentifierSize = keyof (typeof sizeMapping);
const IdentifierSizes = Object.keys(sizeMapping) as IdentifierSize[];

interface Props {
  url?: string | null;
  size?: IdentifierSize;
  circle?: boolean;
}

@Component('Image')
export class Image extends Base<Props> {
  @Prop('image url', { type: String, default: null })
  public url!: string | null;

  @Prop({ acceptableValues: IdentifierSizes, type: String, default: 'm' })
  public size!: IdentifierSize;

  @Prop('is displayed as circle', { type: Boolean, default: false })
  public circle!: boolean;

  public render() {
    return <span class={this.classes} style={this.style}>{this.$slots.default}</span>;
  }

  private get style() {
    const url = this.url;
    if (url == null) { return {}; }
    return {
      'background-image': `url(${this.url})`,
    };
  }

  private get classes() {
    return {
      'fd-image--s': this.size === 's',
      'fd-image--l': this.size === 'l',
      'fd-image--m': this.size === 'm',
      'fd-image--circle': this.circle,
    };
  }
}
