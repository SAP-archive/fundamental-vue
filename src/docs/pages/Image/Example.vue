<title>Size and Shape</title>

<template>
  <div class="fdd-image--default">
    <fd-image size="s" :url="$withBase('images/img-small.jpeg')" />
    <fd-image size="m" :url="$withBase('images/img-small.jpeg')" />
    <fd-image size="l" :url="$withBase('images/img-small.jpeg')" />
    <br /><br />
    <fd-image size="s" circle :url="$withBase('images/img-small.jpeg')" />
    <fd-image size="m" circle :url="$withBase('images/img-small.jpeg')" />
    <fd-image size="l" circle :url="$withBase('images/img-small.jpeg')" />
  </div>
</template>

<style>
.fdd-image--default > span {
  margin-right: 10px;
}
</style>
