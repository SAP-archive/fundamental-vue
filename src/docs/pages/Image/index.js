export default {
  relatedComponents: ["fd-image"]
};
