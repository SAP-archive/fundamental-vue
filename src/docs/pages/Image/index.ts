import { ExampleCollectionFunction } from '../types';

export const plugin: ExampleCollectionFunction = ({}) => {
    return { componentStatus: 'stable', icon: 'background', relatedComponents: [] };
};
