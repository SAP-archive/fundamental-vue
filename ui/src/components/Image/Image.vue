<template>
  <span :class="classes" :style="style">
    <slot />
  </span>
</template>

<script lang="ts">
import Vue from "vue";

const sizeMapping = {
  s: "s",
  m: "m", // default
  l: "l"
};
const IdentifierSizes = Object.keys(sizeMapping);

export default Vue.extend({
  name: "FdImage",
  props: {
    url: { type: String, default: null },
    size: {
      type: String,
      default: sizeMapping.m,
      validator: (value: string) => IdentifierSizes.indexOf(value) >= 0
    },
    circle: { type: Boolean, default: false }
  },
  computed: {
    style(): object {
      const url = this.url;
      if (url == null) {
        return {};
      }
      return {
        backgroundImage: `url(${this.url})`
      };
    },
    classes(): object {
      return {
        [`fd-image--${this.size}`]: true,
        "fd-image--circle": this.circle
      };
    }
  }
});
</script>
