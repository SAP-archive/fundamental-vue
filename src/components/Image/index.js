import Image from "./Image.vue";
import { pluginify } from "./../../util";
export default pluginify(Image);
export { Image };
