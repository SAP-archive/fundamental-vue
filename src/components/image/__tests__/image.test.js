import { mount } from '@vue/test-utils'
import Image from '../Image.vue'

describe('Image', () => {
  it('renders correctly', () => {
    const wrapper = mount(Image, {
      propsData: {
        url: 'https://vuejs.org/images/logo.png',
        size: 'l',
        circle: true
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders correct sizes', () => {
    const imageL = mount(Image, {
      propsData: {
        size: 'l'
      }
    })
    const imageM = mount(Image, {
      propsData: {
        size: 'm'
      }
    })
    const imageS = mount(Image, {
      propsData: {
        size: 's'
      }
    })

    expect(imageL.element).toMatchSnapshot()
    expect(imageM.element).toMatchSnapshot()
    expect(imageS.element).toMatchSnapshot()

    expect(imageL.classes('fd-image--l')).toBe(true)
    expect(imageM.classes('fd-image--m')).toBe(true)
    expect(imageS.classes('fd-image--s')).toBe(true)
  })
})
