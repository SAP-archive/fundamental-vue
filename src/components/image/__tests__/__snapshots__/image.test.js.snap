// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renders correct sizes 1`] = `
<span
  class="fd-image--l"
/>
`;

exports[`Image renders correct sizes 2`] = `
<span
  class="fd-image--m"
/>
`;

exports[`Image renders correct sizes 3`] = `
<span
  class="fd-image--s"
/>
`;

exports[`Image renders correctly 1`] = `
<span
  class="fd-image--l fd-image--circle"
  style="background-image: url(https://vuejs.org/images/logo.png);"
/>
`;
